merkle-tree/hash-set/src/lib.rs

```diff
@@ -431,60 +432,102 @@ where
+        }
+        Err(HashSetError::Full)
+    }
+
+    pub fn find_element(
+        &self,
+        value: &BigUint,
+        current_sequence_number: Option<usize>,
+    ) -> Result<Option<(&mut HashSetCell, I)>, HashSetError> {
         for i in 0..self.capacity_values {
             let probe_index = (value.clone() + i.to_biguint().unwrap() * i.to_biguint().unwrap())
                 % self.capacity_values.to_biguint().unwrap();
             let probe_index = probe_index.to_usize().unwrap();
-            let index_bucket = unsafe { &mut *self.indices.as_ptr().add(probe_index) };
+            let index_bucket = unsafe { &*self.indices.as_ptr().add(probe_index) };
 
             match index_bucket {
-                // The visited hash set cell points to a value in the array.
                 Some(value_index) => {
-                    let value_index =
-                        usize::try_from(*value_index).map_err(|_| HashSetError::UsizeConv)?;
-                    if self.insert_into_occupied_cell(
-                        value_index,
-                        value,
+                    let value_bucket = self.by_value_index(
+                        usize::try_from(*value_index).map_err(|_| HashSetError::UsizeConv)?,
                         current_sequence_number,
-                    )? {
-                        return Ok(());
+                    );
+
+                    if let Some(value_bucket) = value_bucket {
+                        if &value_bucket.value_biguint() == value {
+                            return Ok(Some((value_bucket, *value_index)));
+                        }
                     }
                 }
                 None => {
-                    let value_bucket =
-                        unsafe { &mut *self.values.as_ptr().add(*self.next_value_index) };
-                    // SAFETY: `next_value_index` is always initialized.
-                    *index_bucket = Some(
-                        I::try_from(unsafe { *self.next_value_index })
-                            .map_err(|_| HashSetError::IntegerOverflow)?,
-                    );
-                    *value_bucket = Some(HashSetCell {
-                        value: bigint_to_be_bytes_array(value)?,
-                        sequence_number: None,
-                    });
-                    // SAFETY: `next_value_index` is always initialized.
-                    unsafe {
-                        *self.next_value_index =
-                            if *self.next_value_index < self.capacity_values - 1 {
-                                *self.next_value_index + 1
-                            } else {
-                                0
-                            };
-                    }
-                    return Ok(());
+                    return Ok(None);
                 }
             }
         }
 
-        Err(HashSetError::Full)
+        Ok(None)
     }
 
-    pub fn find_element(
+    /// iterate over a range of elements in the hash set
```

Comment:
Can we improve this docstring and stick to conventions at least a bit?

https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html

* Let's try to use full sententes.
* The preference is to use third person instead of imperative language. "Returns the position" is usually prefered over "Return the position".
* Let's try to not go over 80-100 characters per line (it looks bad on many editors).

We probably need to find some linter for that.